### PR TITLE
Implement QCQP duals

### LINKED ISSUE #309: Error when MOI tries to query the dual solution of a QCQP
I'm running into this error when solving a QCQP with CPLEX. (see below for link to instance).
It seems to be related to the absence/presence of dual variables.

Note that the error occurs within the call to `MOI.optimize!`; calling `CPLEX.optimize!(cpx.inner)` does not raise any error.

```julia
import CPLEX
import MathOptInterface
const MOI = MathOptInterface

mof = MOI.FileFormats.MOF.Model()
MOI.read_from_file(mof, "cgcp.mof")

cpx=  CPLEX.Optimizer()
MOI.copy_to(cpx, mof)

MOI.optimize!(cpx)
```
raises the error

```
Version identifier: 12.10.0.0 | 2019-11-26 | 843d4de
Tried aggregator 1 time.
QCP Presolve eliminated 1 rows and 23 columns.
Aggregator did 14 substitutions.
Reduced QCP has 30 rows, 48 columns, and 85 nonzeros.
Reduced QCP has 11 quadratic constraints.
Presolve time = 0.00 sec. (0.04 ticks)
Parallel mode: using up to 8 threads for barrier.
Number of nonzeros in lower triangle of A*A' = 368
Using Approximate Minimum Degree ordering
Total time for automatic ordering = 0.00 sec. (0.01 ticks)
Summary statistics for Cholesky factor:
  Threads                   = 8
  Rows in Factor            = 30
  Integer space required    = 47
  Total non-zeros in factor = 398
  Total FP ops to factor    = 7066
 Itn      Primal Obj        Dual Obj  Prim Inf Upper Inf  Dual Inf Inf Ratio
   0   5.5683686e+01   0.0000000e+00  2.57e+01  0.00e+00  5.18e+02  1.00e+00
   1   5.3021286e+01  -1.7264892e+00  2.57e+01  0.00e+00  5.18e+02  6.50e-01
   2   7.1670240e+00  -3.3115288e+01  2.55e+01  0.00e+00  5.14e+02  1.05e-01
   3  -6.8434212e+01  -8.0049405e+01  2.26e+01  0.00e+00  4.55e+02  5.75e-02
   4  -2.5242076e+01  -3.3640323e+01  1.31e+01  0.00e+00  2.65e+02  3.01e-01
   5  -6.1799726e+00  -8.1199681e+00  5.30e+00  0.00e+00  1.07e+02  2.78e+00
   6  -3.0068095e+00  -3.9322157e+00  1.04e+00  0.00e+00  2.10e+01  5.31e+00
   7  -1.3315820e+00  -1.7592574e+00  5.04e-01  0.00e+00  1.02e+01  7.54e+00
   8  -7.7572327e-01  -8.9679876e-01  2.54e-01  0.00e+00  5.12e+00  2.32e+01
   9  -4.3256118e-01  -4.5677784e-01  7.43e-02  0.00e+00  1.50e+00  1.09e+02
  10  -4.1508915e-01  -4.2372981e-01  1.51e-02  0.00e+00  3.05e-01  2.96e+02
  11  -3.9575998e-01  -3.9612075e-01  5.44e-03  0.00e+00  1.10e-01  7.05e+03
  12  -3.9503422e-01  -3.9509833e-01  2.27e-04  0.00e+00  4.59e-03  3.94e+04
  13  -3.9496659e-01  -3.9499011e-01  4.05e-05  0.00e+00  8.17e-04  1.03e+05
  14  -3.9491515e-01  -3.9492168e-01  1.50e-05  0.00e+00  3.03e-04  3.57e+05
  15  -3.9490664e-01  -3.9490784e-01  4.23e-06  0.00e+00  8.52e-05  1.88e+06
  16  -3.9490365e-01  -3.9490374e-01  7.81e-07  0.00e+00  1.58e-05  2.33e+07
CPLEX Error  1232: No dual solution exists.

ERROR: CPLEX.CplexError(1232, "CPLEX Error  1232: No dual solution exists.\n")
Stacktrace:
 [1] c_api_getdj at /home/mtanneau/.julia/packages/CPLEX/Hxpuk/src/cpx_solve.jl:150 [inlined]
 [2] optimize!(::CPLEX.Optimizer) at /home/mtanneau/.julia/packages/CPLEX/Hxpuk/src/MOI/MOI_wrapper.jl:1913
 [3] top-level scope at REPL[9]:1
```
(the first line is printed by CPLEX, the rest is Julia's error message).

There seems to be a disagreement between CPLEX and MOI over the dual solution.
MOI claims that the model is solved to optimality, with both primal and dual status at `MOI.FEASIBLE_POINT`.
CPLEX claims there is no dual solution.

Find the problematic instance [here](https://gist.github.com/mtanneau/a8d9dd32f9c6009f175ea7a49f176346), or use the [download link](https://gist.github.com/mtanneau/a8d9dd32f9c6009f175ea7a49f176346/archive/8eb66f88e6dd9f194505b84b62f83550b1636863.zip).

### PR DESCRIPTION
**Pull request by odow**
Closes #295
Closes #309

### CODE REVIEW

**Comment by odow** [comment]
Barring objections, I'll leave this open for another 48 hours, then merge.

**Comment by metab0t** [comment]
I have just noticed that CPLEX.jl is rewritten based on auto-generated bindings with Clang.jl, which makes this package more maintainable. It also makes this PR cleaner. Thanks for your hard work @odow. It's really a new chapter for CPLEX.jl.

**Comment by odow** [comment]
> which makes this package more maintainable. It also makes this PR cleaner

Exactly. It needed to be done. Here's the PR, which contains some discussion you might be interested in reading: https://github.com/jump-dev/CPLEX.jl/pull/316

File: test/MathOptInterface/MOI_wrapper.jl [hunk 1]
```diff
@@ -12,14 +12,10 @@ const CONFIG = MOIT.TestConfig()
 
 const OPTIMIZER = CPLEX.Optimizer()
 MOI.set(OPTIMIZER, MOI.Silent(), true)
-const BRIDGED_OPTIMIZER = MOI.Bridges.full_bridge_optimizer(OPTIMIZER, Float64)
+MOI.set(OPTIMIZER, MOI.RawParameter("CPX_PARAM_REDUCE"), 0)
```
Comment (blegat) [inline test/MathOptInterface/MOI_wrapper.jl:15]: Please comment why we set these

File: src/MOI/MOI_wrapper.jl [hunk 2]
```diff
@@ -2787,15 +2787,61 @@ function MOI.get(
...
+    ind = Vector{Cint}(undef, -surplus_p[])
+    val = Vector{Cdouble}(undef, -surplus_p[])
+    ret = CPXgetqconstrdslack(
+        model.env, model.lp, qind, nz_p, ind, val, -surplus_p[], surplus_p
+    )
+    _check_ret(model, ret)
+    ∇f_max, ∇f_i = findmax(abs.(∇f))
+    for (i, v) in zip(ind, val)
+        if i + 1 == ∇f_i
+            return _dual_multiplier(model) * (∇f_max > cone_top_tol ? v / ∇f[∇f_i] : 0.0)
```
Comment (blegat) [inline src/MOI/MOI_wrapper.jl:2840]: Maybe more readable to do `∇f_max <= cone_top_tol && return 0.0` before the loop and use `dual_multiplier(model) * v / ∇f[∇f_i]` in the loop ?

File: src/MOI/MOI_wrapper.jl [hunk 3]
```diff
@@ -2787,15 +2787,61 @@ function MOI.get(
...
+        model.env, model.lp, qind, nz_p, C_NULL, C_NULL, 0, surplus_p
+    )
+    ind = Vector{Cint}(undef, -surplus_p[])
+    val = Vector{Cdouble}(undef, -surplus_p[])
+    ret = CPXgetqconstrdslack(
+        model.env, model.lp, qind, nz_p, ind, val, -surplus_p[], surplus_p
+    )
+    _check_ret(model, ret)
+    ∇f_max, ∇f_i = findmax(abs.(∇f))
+    for (i, v) in zip(ind, val)
```
Comment (blegat) [inline src/MOI/MOI_wrapper.jl:2838]: So `ind`, `∇f` is like a sparse vector ? In that case, `i + 1 == ∇f_i` is weird as the index in `val` does not match the value of `ind` at the same index.
This loop seems to be checking that there is a nonzero entry in the sparse vector corresponding to the maximal element but it is necessarily the case if it is maximal isn't it ? (Or the vector only contain zeros).
So I would do `_dual_multiplier(model) * (∇f_max > cone_top_tol ? v / ∇f[∇f_i] : 0.0)` directly.

Comment (metab0t) [reply src/MOI/MOI_wrapper.jl:2838]: `ind, val` are sparse vectors and `∇f` is dense vector.
In original CPLEX example, `ind, val` are extended to dense vectors in advance to align with `∇f` to ease iteration.
Here we omit the first step and find the corresponding `∇f_i` directly.
If `∇f_i` is not found in `ind`, the dual is 0.
I agree with your comment below that `∇f_max <= cone_top_tol && return 0.0` should be extracted before loop.

Comment (blegat) [reply src/MOI/MOI_wrapper.jl:2838]: Indeed, I got confused with `val` and `∇f` ^^

**Review by blegat — APPROVED** [review]

**Pull request merged**